src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+
+            // Assign and constrain the final `y_a`.
+            region.assign_fixed(
+                || "qs_2 = 2 on final row",
+                config.q_sinsemilla2,
+                offset - 1,
+                || Ok(pallas::Base::from_u64(2)),
+            )?;
+
+            let y_a_cell = region.assign_advice(
+                || "y_a",
+                config.lambda_1,
+                offset,
+                || y_a.ok_or(Error::SynthesisError),
+            )?;
+
+            // Assign lambda_2 and x_p zero values since they are queried
+            // in the gate.
+            {
+                region.assign_advice(
+                    || "dummy lambda2",
+                    config.lambda_2,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+                region.assign_advice(
+                    || "dummy x_p",
+                    config.x_p,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+            }
+
+            CellValue::new(y_a_cell, y_a.0)
+        };
+
+        #[cfg(test)]
+        #[allow(non_snake_case)]
+        // Check equivalence to result from primitives::sinsemilla::hash_to_point
+        {
+            use crate::circuit::gadget::sinsemilla::message::MessagePiece;
+            use crate::primitives::sinsemilla::{K, S_PERSONALIZATION};
+            use group::prime::PrimeCurveAffine;
+            use pasta_curves::arithmetic::CurveExt;
+
+            let field_elems: Option<Vec<pallas::Base>> =
+                message.iter().map(|piece| piece.field_elem()).collect();
+
+            if field_elems.is_some() {
+                // Get message as a bitstring.
+                let bitstring: Vec<bool> = message
+                    .iter()
+                    .map(|piece: &MessagePiece<pallas::Base, K>| {
+                        piece
+                            .field_elem()
+                            .unwrap()
+                            .to_le_bits()
+                            .into_iter()
+                            .take(K * piece.num_words())
+                            .collect::<Vec<_>>()
+                    })
+                    .flatten()
+                    .collect();
+
+                let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+                let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+                let expected_point = bitstring
```

Comment:
```suggestion
                // We can use complete addition here because it differs from
                // incomplete addition with negligible probability.
                let expected_point = bitstring
```